File: JANL/UserControls/MonthPicker.vb
Option Strict On

Imports System.ComponentModel
Imports System.Globalization

<Designer(GetType(HResizeOnly))>
Public Class MonthPicker
	Private _MinYear As Integer
	Private Months As List(Of MonthItem)

	Public Sub New()
		InitializeComponent()

		Dim T = Date.Today
		Months = Enumerable.Range(1, 12).Select(Function(i) New MonthItem(CUShort(i), DateTimeFormatInfo.CurrentInfo.GetMonthName(i))).ToList()
		CB_Month.DataSource = Months
		CB_Month.DisplayMember = "Name"
		CB_Month.ValueMember = "Index"
		CB_Month.SelectedValue = CUShort(T.Month)

		_MinYear = 2010
		FillYear()
	End Sub

#Region "Properties"

#Region "Designer"

	''' <summary>
	''' Минимальный год
	''' </summary>
	<Browsable(True), Category("MonthPicker"), DefaultValue(2010)>
	Public Property MinYear As Integer
		Get
			Return _MinYear
		End Get
		Set(value As Integer)
			_MinYear = value
			FillYear()
		End Set
	End Property

#End Region

	''' <summary>
	''' Первый день месяца
	''' </summary>
	<Browsable(True), Category("MonthPicker")>
	Public ReadOnly Property FirstDate As Date
		Get
			Return New Date(Year, Month, 1)
		End Get
	End Property

	''' <summary>
	''' Последний день месяца
	''' </summary>
	<Browsable(True), Category("MonthPicker")>
	Public ReadOnly Property LastDate As Date
		Get
			Return New Date(Year, Month, Date.DaysInMonth(Year, Month))
		End Get
	End Property

	''' <summary>
	''' Выбранный месяц
	''' </summary>
	<Browsable(False)>
	Public ReadOnly Property Month As UShort
		Get
			Return CUShort(CB_Month.SelectedValue)
		End Get
	End Property

	''' <summary>
	''' Выбранный год
	''' </summary>
	<Browsable(False)>
	Public ReadOnly Property Year As Integer
		Get
			Return CInt(CB_Year.SelectedItem)
		End Get
	End Property

#End Region

	Protected Sub OnDateChanged()
		RaiseEvent DateChanged(Me, EventArgs.Empty)
	End Sub

	Private Sub CB_Month_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CB_Month.SelectedIndexChanged
		OnDateChanged()
	End Sub

	Private Sub CB_Year_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CB_Year.SelectedIndexChanged
		OnDateChanged()
	End Sub

	Private Sub FillYear()
		Dim Y = Date.Today.Year
		CB_Year.DataSource = Enumerable.Range(MinYear, Y - MinYear + 1).ToList()
		If MinYear <= Y Then CB_Year.Text = Y.ToString
	End Sub

	Private Structure MonthItem

		Public Sub New(I As UShort, N As String)
			Index = I
			Name = N
		End Sub

		Public Property Index As UShort
		Public Property Name As String
	End Structure

	Public Event DateChanged(sender As Object, e As EventArgs)

End Class

File: JANL.WinForms/UserControls/MonthPicker.cs
using JANL.Designers;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Windows.Forms;

namespace JANL.UserControls
{
    /// <summary>
    /// Элемент выбора месяца
    /// </summary>
    [Designer(typeof(HResizeOnly))]
    public partial class MonthPicker : UserControl
    {
        private readonly List<MonthItem> Months;

        /// <summary>
        /// Создаёт новы элемент выбора месяца
        /// </summary>
        public MonthPicker()
        {
            InitializeComponent();

            var T = DateTime.Today;
            Months = Enumerable.Range(1, 12).Select(i => new MonthItem(i)).ToList();
            CB_Month.DataSource = Months;
            CB_Month.DisplayMember = "Name";
            CB_Month.ValueMember = "Index";
            CB_Month.SelectedValue = T.Month;

            _MinYear = 2010;
            FillYear();
        }

        /// <summary>
        /// Вызывает <see cref="DateChanged"/>
        /// </summary>
        protected void OnDateChanged() => DateChanged?.Invoke(this, EventArgs.Empty);

        private void CB_Month_SelectedIndexChanged(object sender, EventArgs e) => OnDateChanged();

        private void CB_Year_SelectedIndexChanged(object sender, EventArgs e) => OnDateChanged();

        private void FillYear()
        {
            var Y = MaxYear ?? DateTime.Today.Year;
            CB_Year.DataSource = Enumerable.Range(MinYear, Y - MinYear + 1).ToList();
            if (MinYear <= Y) { CB_Year.Text = Y.ToString(); }
        }

        #region Properties
        private int? _MaxYear;
        private int _MinYear;

        /// <summary>
        /// Первый день месяца
        /// </summary>
        [Browsable(true), Category("MonthPicker")]
        public DateTime FirstDate => new DateTime(Year, Month, 1);

        /// <summary>
        /// Последний день месяца
        /// </summary>
        [Browsable(true), Category("MonthPicker")]
        public DateTime LastDate => new DateTime(Year, Month, DateTime.DaysInMonth(Year, Month));

        /// <summary>
        /// Максимальный год
        /// </summary>
        [Browsable(true), Category("MonthPicker"), DefaultValue(null), Description("Максимальный год")]
        public int? MaxYear
        {
            get => _MaxYear;
            set
            {
                _MaxYear = value;
                FillYear();
            }
        }

        /// <summary>
        /// Минимальный год
        /// </summary>
        [Browsable(true), Category("MonthPicker"), DefaultValue(2010), Description("Минимальный год")]
        public int MinYear
        {
            get => _MinYear;
            set
            {
                _MinYear = value;
                FillYear();
            }
        }

        /// <summary>
        /// Выбранный месяц
        /// </summary>
        [Browsable(false)]
        public int Month => (int)CB_Month.SelectedValue;

        /// <summary>
        /// Выбранный год
        /// </summary>
        [Browsable(false)]
        public int Year => (int)CB_Year.SelectedItem;

        #endregion Properties

        private struct MonthItem : IEquatable<MonthItem>
        {
            public MonthItem(int i)
            {
                Index = i;
                Name = $"{i:00} - {DateTimeFormatInfo.CurrentInfo.GetMonthName(i)}";
            }

            public int Index { get; set; }
            public string Name { get; set; }

            public bool Equals(MonthItem other) => Index.Equals(other);
        }

        /// <summary>
        /// Происходит при смене месяца
        /// </summary>
        public event EventHandler DateChanged;
    }
}